core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/sync/BufferingContentStreamProvider.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.sync;
+
+import static software.amazon.awssdk.utils.FunctionalUtils.invokeSafely;
+
+import java.io.BufferedInputStream;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import software.amazon.awssdk.annotations.NotThreadSafe;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.io.ReleasableInputStream;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.utils.IoUtils;
+
+/**
+ * {@code ContentStreamProvider} implementation that buffers the data stream data to memory as it's read. Once the underlying
+ * stream is read fully, all subsequent calls to {@link #newStream()} will use the buffered data.
+ */
+@SdkInternalApi
+@NotThreadSafe
+public final class BufferingContentStreamProvider implements ContentStreamProvider {
+    private final ContentStreamProvider delegate;
+    private InputStream bufferedStream;
+
+    private byte[] bufferedStreamData;
+    private int count;
+
+    public BufferingContentStreamProvider(ContentStreamProvider delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public InputStream newStream() {
+        if (bufferedStreamData != null) {
+            return new ByteArrayInputStream(bufferedStreamData, 0, this.count);
+        }
+
+        if (bufferedStream == null) {
+            InputStream delegateStream = delegate.newStream();
+            bufferedStream = new BufferStream(delegateStream);
+            IoUtils.markStreamWithMaxReadLimit(bufferedStream, Integer.MAX_VALUE);
+
+            // Make non-closeable so that the SDK can't close those prematurely, before we can buffer all the data.
+            bufferedStream = ReleasableInputStream.wrap(bufferedStream).disableClose();
```

Comment:
When would we close the delegate stream, then?

Reply:
Unfortunately, nothing. I don't think it's possible given that we can't rely on the SDK to be close previous streams for us.

Reply:
How did it work before?

Reply:
The contract is that the implementation needs to close the previous stream when `newStream()` is called. The SDK will close the final stream after the request is sent.

Reply:
https://github.com/aws/aws-sdk-java-v2/blob/b1b5216bf2d8558f9f638152d4c2bfff4289eebf/http-client-spi/src/main/java/software/amazon/awssdk/http/ContentStreamProvider.java#L34-L37

Reply:
Update, discussed offline, we changed to remove non-closing.